src/game.cpp

```diff
@@ -6494,60 +6497,87 @@ void game::explosion( const tripoint &p, float power, float factor,
+                distrib[ e.first ] = std::make_pair( -1, e.second );
+            }
+        }
     }
+
+    return distrib;
 }
 
-void game::shrapnel( const tripoint &p, int power, int count, int radius )
+std::unordered_map<tripoint,int> game::shrapnel( const tripoint &src, int power, int count, int mass, int range )
 {
-    if( power <= 0 ) {
-        return;
+    if( range < 0 ) {
+        range = std::max( ( 2 * log( power / 2  ) ) + 2, 0.0 );
     }
 
-    if( radius < 0 ) {
-        return;
-    }
+    // contains of all tiles considered with value being sum of damage received (if any)
+    std::unordered_map<tripoint,int> distrib;
 
-    npc fake_npc;
-    fake_npc.name = _("Shrapnel");
-    fake_npc.set_fake(true);
-    fake_npc.setpos( p );
     projectile proj;
-    proj.speed = 100;
-    proj.range = radius;
-    proj.proj_effects.insert( "DRAW_AS_LINE" );
+    proj.speed = 1000; // no dodging shrapnel
+    proj.range = range;
     proj.proj_effects.insert( "NULL_SOURCE" );
-    for( int i = 0; i < count; i++ ) {
-        // TODO: Z-level shrapnel, but not before z-level ranged attacks
-        tripoint sp{ static_cast<int> (rng( p.x - radius, p.x + radius )),
-                     static_cast<int> (rng( p.y - radius, p.y + radius )),
-                     p.z };
 
-        proj.impact = damage_instance::physical( power, power, 0, 0 );
+    for( auto i = 0; i != count; ++i ) {
+        // shrapnel expands randomly in all directions
+        tripoint dst = m.random_perimeter( src, range );
 
-        Creature *critter_in_center = critter_at( p ); // Very unfortunate critter
-        if( critter_in_center != nullptr ) {
-            dealt_projectile_attack dda; // Cool variable name
-            dda.proj = proj;
-            // For first shrapnel piece:
-            // 50% chance for 50%-100% base (power to 2 * power)
-            // 50% chance for 0-25% base
-            // Each one after that gets a progressively lower chance of hitting
-            dda.missed_by = rng_float( 0.4, 1.0 ) + (i * 1.0 / count);
-            critter_in_center->deal_projectile_attack( nullptr, dda );
-        }
+        int kinetic = power;
+        bresenham( src, dst, 0, 0, [this,&distrib,&i,&src,&kinetic,&mass,&proj]( const tripoint& e ) {
 
-        if( sp != p ) {
-            // This needs to be high enough to prevent game from thinking that
-            //  the fake npc is scoring headshots.
-            fake_npc.projectile_attack( proj, sp, 3600 );
-        }
+            distrib[ e ] += 0; // add this tile to the distribution
+
+            auto critter = critter_at( e );
+            if( critter && !critter->is_dead_state() ) {
+                // special case critter at epicenter to have equivalent chance to adjacent tile
+                if( src == e && !one_in( 8 ) ) {
+                    return true;
+                }
+
+                dealt_projectile_attack frag;
+                frag.proj = proj;
+                frag.missed_by = 0;
```

Comment:
Headshots only ever target the head and have significantly higher damage (2.45 to 3.35 of the base damage).

A `missed_by` value between 0.6 and 0.2 would be better - it would result in "Good hit"s and regular (no message) hits, which deal 1.0-1.5 and 0.5-1.0 of base damage respectively. They would also target torso and limbs instead of the head. Currently there is no way to have consistent ranged damage and both torso and head hits.